stateproof/builder.go

```diff
@@ -20,48 +20,163 @@ import (
 	"database/sql"
 	"encoding/binary"
+	"errors"
 	"fmt"
 	"sort"
 
 	"github.com/algorand/go-algorand/config"
 	"github.com/algorand/go-algorand/crypto/stateproof"
 	"github.com/algorand/go-algorand/data/basics"
-	"github.com/algorand/go-algorand/data/bookkeeping"
 	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
 	"github.com/algorand/go-algorand/logging"
 	"github.com/algorand/go-algorand/network"
 	"github.com/algorand/go-algorand/protocol"
 	"github.com/algorand/go-algorand/stateproof/verify"
 )
 
-// makeBuilderForRound not threadsafe, should be called in a lock environment
-func (spw *Worker) makeBuilderForRound(rnd basics.Round) (builder, error) {
-	l := spw.ledger
-	hdr, err := l.BlockHdr(rnd)
+var errVotersNotTracked = errors.New("voters not tracked for the given lookback round")
+
+// OnPrepareVoterCommit is a function called by the voters tracker when it's preparing to commit rnd. It gives the builder
+// the chance to persist the data it needs.
+func (spw *Worker) OnPrepareVoterCommit(oldBase basics.Round, newBase basics.Round, votersFetcher ledgercore.LedgerForSPBuilder) {
+	for rnd := oldBase + 1; rnd <= newBase; rnd++ {
+		header, err := votersFetcher.BlockHdr(rnd)
+		if err != nil {
+			spw.log.Errorf("OnPrepareVoterCommit(%d): could not fetch round header: %v", rnd, err)
+			continue
+		}
+
+		proto := config.Consensus[header.CurrentProtocol]
+		if proto.StateProofInterval == 0 || uint64(rnd)%proto.StateProofInterval != 0 {
+			continue
+		}
+
+		builderExists, err := spw.builderExists(rnd)
+		if err != nil {
+			spw.log.Warnf("OnPrepareVoterCommit(%d): could not check builder existence, assuming it doesn't exist: %v\n", rnd, err)
+		} else if builderExists {
+			continue
+		}
+
+		buildr, err := createBuilder(rnd, votersFetcher)
+		if err != nil {
+			if errors.Is(err, errVotersNotTracked) {
+				// Voters not tracked for that round.  Might not be a valid
+				// state proof round; state proofs might not be enabled; etc.
+				spw.log.Warnf("OnPrepareVoterCommit(%d): %v", rnd, err)
+				continue
+			}
+
+			spw.log.Errorf("OnPrepareVoterCommit(%d): could not create builder: %v", rnd, err)
+			continue
+		}
+
+		err = spw.db.Atomic(func(_ context.Context, tx *sql.Tx) error {
+			return persistBuilder(tx, rnd, &buildr)
+		})
+		if err != nil {
+			spw.log.Errorf("OnPrepareVoterCommit(%d): could not persist builder: %v", rnd, err)
+		}
+	}
+}
+
+// loadOrCreateBuilderWithSignatures either loads a builder from the DB or creates a new builder.
+// this function fills the builder with all the available signatures
+func (spw *Worker) loadOrCreateBuilderWithSignatures(rnd basics.Round) (builder, error) {
+	b, err := spw.loadOrCreateBuilder(rnd)
+	if err != nil {
+		return builder{}, err
+	}
+
+	if err := spw.loadSignaturesIntoBuilder(&b); err != nil {
+		return builder{}, err
+	}
+	return b, nil
+}
+
+func (spw *Worker) loadOrCreateBuilder(rnd basics.Round) (builder, error) {
```

Comment:
No need to make these 2 functions. `loadOrCreateBuilder` is never used outside of `loadOrCreateBuilderWithSignatures`.

Reply:
In this case, we use functions for readability. Both `loadOrCreateBuilder` and `loadSignaturesIntoBuilder` are pretty long. Combine them together and we will get one huge function. With the current approach, A reader can understand the two high-level operation that needs to be done. 1 - create(or load) builder. 2 - load signatures.